### PR TITLE
[wearable_rotary] Add Wear OS support

### PR DESCRIPTION
**Pull request by Rexios80**
- Adds Wear OS support for rotary events
- Breaking: `RotaryEvent` is now a class with `direction` and `magnitude` properties
- Adds `RotaryScrollController` for easily making scrollable widgets respond to rotary events

### CODE REVIEW

**Comment by swift-kim** [comment]
Thank you for your contribution.

@bbrto21 @HakkyuKim How do you like adding a non-Tizen project to this repo? Could you test the code in this change if you have a Galaxy Watch 4 or any other Wear OS device?

**Comment by bbrto21** [comment]
@swift-kim 
>@bbrto21 @HakkyuKim How do you like adding a non-Tizen project to this repo? 

Well... I can't see any particular reason not to add this and I think developers who want to expand this are going to try to contribute to this naturally because this plugin is original one.

>  Could you test the code in this change if you have a Galaxy Watch 4 or any other Wear OS device?

I can test it with my Galaxy Watch 4 when I'm working from home next week.

**Comment by swift-kim** [comment]
@bbrto21 My concern is that we don't have a CI testbed set up for Wear OS for testing the code, and not many of us have enough experience with Android/Kotlin to maintain the code. If you still don't mind, please go ahead.

**Comment by bbrto21** [comment]
@Rexios80 This PR causes the view to bounce when scrolling is no longer possible. Is it intentional? or Is this the normal behavior of the scroll controller??

**original**

**This PR**

**Comment by bbrto21** [comment]
> @bbrto21 My concern is that we don't have a CI testbed set up for Wear OS for testing the code, and not many of us have enough experience with Android/Kotlin to maintain the code. If you still don't mind, please go ahead.

@swift-kim You are right. we still don't have a good testing strategy for some of our plugins. this thorny problem should not be increased.

@Rexios80 Since the situation is like this, I would like to ask you if you have a strategy to test this and a plan to develop it.

**Comment by Rexios80** [comment]
I think I fixed the overscroll behavior.

If we don't want this code in this repo, then we need to make a platform interface so that the Wear OS code can exist somewhere else but still be used with this plugin. I would rather not do that as that would make this plugin significantly more complicated for how simple it is.

My plan to test it is to use it in my apps. If anyone has issues with the Wear OS code, they can submit a PR just like I did. If the code lived somewhere else, it would be the same.

**Comment by bbrto21** [comment]
Looks good!

**Comment by bbrto21** [comment]
@HakkyuKim  @swift-kim As you can see it works fine on the emulator. 

I'm not familiar with Kotlin and Android, but...I think this is fine overall. what do you think?
And Should the Samsung license notice be added to the newly added code?

**Comment by HakkyuKim** [comment]
I think it's great to have this in our repo, it could be a good example of a plugin that supports both Tizen and Android. I do wish to test this with CI though I'm not sure if rotation can be tested programatically.

**Comment by bbrto21** [comment]
@swift-kim @HakkyuKim 
This patch has breaking changes on dart side. what do you think? this change looks good to me.

File: packages/wearable_rotary/example/android/app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -0,0 +1,40 @@
+<manifest xmlns:android="http://schemas.android.com/apk/res/android"
+    package="com.samsung.wearable_rotary_example">
+
+   <uses-feature android:name="android.hardware.type.watch" />
+
+   <application
+        android:label="wearable_rotary_example"
```
[the remark this answers is not in the document]
Comment (swift-kim) [reply packages/wearable_rotary/example/android/app/src/main/AndroidManifest.xml:7]: Uhm.. why?

[the remark this answers is not in the document]
Comment (Rexios80) [reply packages/wearable_rotary/example/android/app/src/main/AndroidManifest.xml:7]: Yeah these are all generated by `flutter create` and probably shouldn't be touched

File: packages/wearable_rotary/README.md [hunk 2]
```diff
@@ -2,15 +2,35 @@
...
+
+```kotlin
+import android.view.MotionEvent
+import com.samsung.wearable_rotary.WearableRotaryPlugin
+
+class MainActivity : FlutterActivity() {
+    override fun onGenericMotionEvent(event: MotionEvent?): Boolean {
+        return when {
+            WearableRotaryPlugin.onGenericMotionEvent(event) -> true
+            else -> super.onGenericMotionEvent(event)
```
[the remark this answers is not in the document]
Comment (Rexios80) [reply packages/wearable_rotary/README.md:21]: I tried but couldn't figure out how to hook into `onGenericMotionEvent` without breaking Flutter's own gesture handling

File: packages/wearable_rotary/example/android/app/build.gradle [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    kotlinOptions {
+        jvmTarget = '1.8'
+    }
+
+    sourceSets {
+        main.java.srcDirs += 'src/main/kotlin'
+    }
+
+    defaultConfig {
+        // TODO: Specify your own unique Application ID (https://developer.android.com/studio/build/application-id.html).
```
[the remark this answers is not in the document]
Comment (bbrto21) [reply packages/wearable_rotary/example/android/app/build.gradle:46]: Is this also an auto-generated comment? Even automatically generated comments, we try to remove as much as possible when they are useless.

[the remark this answers is not in the document]
Comment (Rexios80) [reply packages/wearable_rotary/example/android/app/build.gradle:46]: Most everything in the example/android folder is generated by `flutter create`

**Review by bbrto21 — APPROVED** [review]
Generally LGTM.
But I'm not familiar enough with Android, so I'd love to have someone else's additional approval.

File: packages/wearable_rotary/example/android/app/build.gradle (hunk 3 above)
[the remark this answers is not in the document]
Comment (bbrto21) [reply packages/wearable_rotary/example/android/app/build.gradle:46]: thank you for quick answer :)

**Review by swift-kim — COMMENTED** [review]
Otherwise looks good to me.

**Review by HakkyuKim — COMMENTED** [review]
We could reference the Android build script from the flutter plugins repo for CI in the future. I'm Okay with the current changes.

File: packages/wearable_rotary/android/src/main/kotlin/com/samsung/wearable_rotary/WearableRotaryPlugin.kt [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+            EventChannel(flutterPluginBinding.binaryMessenger, "flutter.wearable_rotary.channel")
+        channel.setStreamHandler(this)
+
+        val context = flutterPluginBinding.applicationContext
+        scaleFactor = ViewConfigurationCompat.getScaledVerticalScrollFactor(
+            ViewConfiguration.get(context), context
+        )
+    }
+
+    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {}
```
[the remark this answers is not in the document]
Comment (Rexios80) [reply packages/wearable_rotary/android/src/main/kotlin/com/samsung/wearable_rotary/WearableRotaryPlugin.kt:50]: There was a reason I didn't put it there, but I can't remember what is it so now it's there

File: packages/wearable_rotary/android/src/main/kotlin/com/samsung/wearable_rotary/WearableRotaryPlugin.kt [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+            val delta = event.getAxisValue(MotionEventCompat.AXIS_SCROLL) * scaleFactor
+            events?.success(
+                mapOf(
+                    "direction" to if (delta < 0) "clockwise" else "counterClockwise",
+                    "magnitude" to abs(delta),
+                )
+            )
+            return true
+
```
[the remark this answers is not in the document]
Comment (Rexios80) [reply packages/wearable_rotary/android/src/main/kotlin/com/samsung/wearable_rotary/WearableRotaryPlugin.kt:35]: Done

File: packages/wearable_rotary/android/src/main/kotlin/com/samsung/wearable_rotary/WearableRotaryPlugin.kt [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        fun onGenericMotionEvent(event: MotionEvent?): Boolean {
+            if (event == null || event.action != MotionEvent.ACTION_SCROLL || !event.isFromSource(
+                    InputDeviceCompat.SOURCE_ROTARY_ENCODER
+                )
+            ) {
+                return false
+            }
+
+            val delta = event.getAxisValue(MotionEventCompat.AXIS_SCROLL) * scaleFactor
```
[the remark this answers is not in the document]
Comment (Rexios80) [reply packages/wearable_rotary/android/src/main/kotlin/com/samsung/wearable_rotary/WearableRotaryPlugin.kt:27]: The if statement above checks if `event` is null

**Review by swift-kim — APPROVED** [review]
Thank you senpai <3

**Pull request merged**